File: Leetcode/0025-reverse-nodes-in-k-group/solution.go
package main

import (
	"fmt"
	"reflect"
)

// ============================================================
// 0025. Reverse Nodes in k-Group
// https://leetcode.com/problems/reverse-nodes-in-k-group/
// Difficulty: Hard
// Tags: Linked List, Recursion
// ============================================================

// ListNode — singly-linked list node
type ListNode struct {
	Val  int
	Next *ListNode
}

// reverseKGroup — Approach 1: Iterative
// Process list in groups of k, reversing each group in-place.
// Time:  O(n) — each node is visited at most twice
// Space: O(1) — only constant extra pointers
func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{Next: head}
	groupPrev := dummy

	for {
		// Check if k nodes remain
		kth := getKthNode(groupPrev, k)
		if kth == nil {
			break
		}

		groupNext := kth.Next

		// Reverse k nodes in this group
		prev := kth.Next
		curr := groupPrev.Next
		for curr != groupNext {
			tmp := curr.Next
			curr.Next = prev
			prev = curr
			curr = tmp
		}

		// Reconnect
		tmp := groupPrev.Next // original first node, now the tail
		groupPrev.Next = kth  // point to new head of reversed group
		groupPrev = tmp       // advance to the tail of the reversed group
	}

	return dummy.Next
}

// getKthNode returns the kth node after the given node, or nil if fewer than k nodes remain.
func getKthNode(node *ListNode, k int) *ListNode {
	for node != nil && k > 0 {
		node = node.Next
		k--
	}
	return node
}

// reverseKGroupRecursive — Approach 2: Recursive
// Reverse first k nodes, recurse on the rest, connect them.
// Time:  O(n) — each node is visited at most twice
// Space: O(n/k) — recursion stack depth equals number of groups
func reverseKGroupRecursive(head *ListNode, k int) *ListNode {
	// Check if k nodes exist
	node := head
	count := 0
	for node != nil && count < k {
		node = node.Next
		count++
	}

	if count < k {
		return head // not enough nodes, leave as is
	}

	// Reverse first k nodes
	var prev *ListNode
	curr := head
	for i := 0; i < k; i++ {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	// head is now the tail of the reversed group — connect to recursion result
	head.Next = reverseKGroupRecursive(curr, k)

	return prev // prev is the new head of the reversed group
}

// ============================================================
// Helper Functions for Testing
// ============================================================

// buildList creates a linked list from a slice of integers
func buildList(nums []int) *ListNode {
	dummy := &ListNode{}
	curr := dummy
	for _, v := range nums {
		curr.Next = &ListNode{Val: v}
		curr = curr.Next
	}
	return dummy.Next
}

// listToSlice converts a linked list back to a slice for easy comparison
func listToSlice(head *ListNode) []int {
	result := []int{}
	for head != nil {
		result = append(result, head.Val)
		head = head.Next
	}
	return result
}

// ============================================================
// Test Cases
// ============================================================

func main() {
	passed, failed := 0, 0
	test := func(name string, got, expected []int) {
		if reflect.DeepEqual(got, expected) {
			fmt.Printf("  ✅ PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("  ❌ FAIL: %s\n    Got:      %v\n    Expected: %v\n", name, got, expected)
			failed++
		}
	}

	// --- Iterative approach tests ---
	fmt.Println("Approach 1: Iterative")

	test("[1,2,3,4,5] k=2 -> [2,1,4,3,5]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3, 4, 5}), 2)),
		[]int{2, 1, 4, 3, 5})

	test("[1,2,3,4,5] k=3 -> [3,2,1,4,5]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3, 4, 5}), 3)),
		[]int{3, 2, 1, 4, 5})

	test("[1,2,3,4,5] k=1 -> [1,2,3,4,5]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3, 4, 5}), 1)),
		[]int{1, 2, 3, 4, 5})

	test("[1] k=1 -> [1]",
		listToSlice(reverseKGroup(buildList([]int{1}), 1)),
		[]int{1})

	test("[1,2] k=2 -> [2,1]",
		listToSlice(reverseKGroup(buildList([]int{1, 2}), 2)),
		[]int{2, 1})

	test("[1,2,3,4] k=2 -> [2,1,4,3]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3, 4}), 2)),
		[]int{2, 1, 4, 3})

	test("[1,2,3,4,5] k=5 -> [5,4,3,2,1]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3, 4, 5}), 5)),
		[]int{5, 4, 3, 2, 1})

	test("[1,2,3] k=3 -> [3,2,1]",
		listToSlice(reverseKGroup(buildList([]int{1, 2, 3}), 3)),
		[]int{3, 2, 1})

	// --- Recursive approach tests ---
	fmt.Println("\nApproach 2: Recursive")

	test("[1,2,3,4,5] k=2 -> [2,1,4,3,5]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1, 2, 3, 4, 5}), 2)),
		[]int{2, 1, 4, 3, 5})

	test("[1,2,3,4,5] k=3 -> [3,2,1,4,5]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1, 2, 3, 4, 5}), 3)),
		[]int{3, 2, 1, 4, 5})

	test("[1,2,3,4,5] k=1 -> [1,2,3,4,5]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1, 2, 3, 4, 5}), 1)),
		[]int{1, 2, 3, 4, 5})

	test("[1] k=1 -> [1]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1}), 1)),
		[]int{1})

	test("[1,2,3,4,5] k=5 -> [5,4,3,2,1]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1, 2, 3, 4, 5}), 5)),
		[]int{5, 4, 3, 2, 1})

	test("[1,2,3,4] k=2 -> [2,1,4,3]",
		listToSlice(reverseKGroupRecursive(buildList([]int{1, 2, 3, 4}), 2)),
		[]int{2, 1, 4, 3})

	// Results
	fmt.Printf("\n📊 Results: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0025-reverse-nodes-in-k-group/solution.py
# ============================================================
# 0025. Reverse Nodes in k-Group
# https://leetcode.com/problems/reverse-nodes-in-k-group/
# Difficulty: Hard
# Tags: Linked List, Recursion
# ============================================================

from __future__ import annotations


class ListNode:
    """Singly-linked list node."""
    def __init__(self, val: int = 0, next: ListNode | None = None):
        self.val = val
        self.next = next


class Solution:
    def reverseKGroup(self, head: ListNode | None, k: int) -> ListNode | None:
        """
        Approach 1: Iterative
        Process list in groups of k, reversing each group in-place.
        Time:  O(n) — each node is visited at most twice
        Space: O(1) — only constant extra pointers
        """
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            # Check if k nodes remain
            kth = self._getKthNode(groupPrev, k)
            if not kth:
                break

            groupNext = kth.next

            # Reverse k nodes in this group
            prev, curr = kth.next, groupPrev.next
            while curr != groupNext:
                tmp = curr.next
                curr.next = prev
                prev = curr
                curr = tmp

            # Reconnect: groupPrev -> new head (kth), old head -> groupPrev for next iteration
            tmp = groupPrev.next   # original first node, now the tail
            groupPrev.next = kth   # point to new head of reversed group
            groupPrev = tmp        # advance to the tail of the reversed group

        return dummy.next

    def _getKthNode(self, node: ListNode | None, k: int) -> ListNode | None:
        """Returns the kth node after the given node, or None if fewer than k nodes remain."""
        while node and k > 0:
            node = node.next
            k -= 1
        return node

    def reverseKGroupRecursive(self, head: ListNode | None, k: int) -> ListNode | None:
        """
        Approach 2: Recursive
        Reverse first k nodes, recurse on the rest, connect them.
        Time:  O(n) — each node is visited at most twice
        Space: O(n/k) — recursion stack depth equals number of groups
        """
        # Check if k nodes exist
        node = head
        count = 0
        while node and count < k:
            node = node.next
            count += 1

        if count < k:
            return head  # not enough nodes, leave as is

        # Reverse first k nodes
        prev, curr = None, head
        for _ in range(k):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        # head is now the tail of the reversed group — connect to recursion result
        head.next = self.reverseKGroupRecursive(curr, k)

        return prev  # prev is the new head of the reversed group


# ============================================================
# Helper Functions for Testing
# ============================================================

def build_list(nums: list[int]) -> ListNode | None:
    """Creates a linked list from a list of integers."""
    dummy = ListNode()
    curr = dummy
    for v in nums:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def list_to_array(head: ListNode | None) -> list[int]:
    """Converts a linked list back to a Python list for easy comparison."""
    result = []
    while head is not None:
        result.append(head.val)
        head = head.next
    return result


# ============================================================
# Test Cases
# ============================================================

if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0

    def test(name: str, got, expected):
        global passed, failed
        if got == expected:
            print(f"  \u2705 PASS: {name}")
            passed += 1
        else:
            print(f"  \u274c FAIL: {name}")
            print(f"    Got:      {got}")
            print(f"    Expected: {expected}")
            failed += 1

    # --- Iterative approach tests ---
    print("Approach 1: Iterative")

    test("[1,2,3,4,5] k=2 -> [2,1,4,3,5]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3, 4, 5]), 2)),
         [2, 1, 4, 3, 5])

    test("[1,2,3,4,5] k=3 -> [3,2,1,4,5]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3, 4, 5]), 3)),
         [3, 2, 1, 4, 5])

    test("[1,2,3,4,5] k=1 -> [1,2,3,4,5]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3, 4, 5]), 1)),
         [1, 2, 3, 4, 5])

    test("[1] k=1 -> [1]",
         list_to_array(sol.reverseKGroup(build_list([1]), 1)),
         [1])

    test("[1,2] k=2 -> [2,1]",
         list_to_array(sol.reverseKGroup(build_list([1, 2]), 2)),
         [2, 1])

    test("[1,2,3,4] k=2 -> [2,1,4,3]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3, 4]), 2)),
         [2, 1, 4, 3])

    test("[1,2,3,4,5] k=5 -> [5,4,3,2,1]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3, 4, 5]), 5)),
         [5, 4, 3, 2, 1])

    test("[1,2,3] k=3 -> [3,2,1]",
         list_to_array(sol.reverseKGroup(build_list([1, 2, 3]), 3)),
         [3, 2, 1])

    # --- Recursive approach tests ---
    print("\nApproach 2: Recursive")

    test("[1,2,3,4,5] k=2 -> [2,1,4,3,5]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1, 2, 3, 4, 5]), 2)),
         [2, 1, 4, 3, 5])

    test("[1,2,3,4,5] k=3 -> [3,2,1,4,5]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1, 2, 3, 4, 5]), 3)),
         [3, 2, 1, 4, 5])

    test("[1,2,3,4,5] k=1 -> [1,2,3,4,5]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1, 2, 3, 4, 5]), 1)),
         [1, 2, 3, 4, 5])

    test("[1] k=1 -> [1]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1]), 1)),
         [1])

    test("[1,2,3,4,5] k=5 -> [5,4,3,2,1]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1, 2, 3, 4, 5]), 5)),
         [5, 4, 3, 2, 1])

    test("[1,2,3,4] k=2 -> [2,1,4,3]",
         list_to_array(sol.reverseKGroupRecursive(build_list([1, 2, 3, 4]), 2)),
         [2, 1, 4, 3])

    print(f"\n\U0001f4ca Results: {passed} passed, {failed} failed, {passed + failed} total")
